Add the initial capability to checkpoint partial server state

e-kayrakli opened the pull request:
This PR adds the ability to checkpoint pdarrays on the server side. Core of this implementation is from @bmcdonald3.

Two client-side functions are added:

- `arkouda.save_checkpoint(name='', path='.akdata')`
  - `path` is the directory to store the save instance. It will be created if it doesn't exist.
  - `name` is the name of the save instance, by default this will be the server id
- `arkouda.load_checkpoint(name)`
  - loads the checkpoint with the given name

TODO
- [x] handle merge conflicts
- [x] add test
- [x] test with large number of nodes and arrays

ajpotts commented on src/ParquetMsg.chpl:
```diff
@@ -1147,18 +1168,22 @@ module ParquetMsg {
+                                           compression:int, mode, e.a)[0];
       } otherwise {
         var errorMsg = "Writing Parquet files not supported for %s type".format(msgArgs.getValueOf("dtype"));
         pqLogger.error(getModuleName(),getRoutineName(),getLineNumber(),errorMsg);
```
I wasn't able to make the PR so that I could try it out:
```


$ make
/home/amandapotts/git/arkouda//src/registry/register_commands.bash /home/amandapotts/git/arkouda//src/registry /home/amandapotts/git/arkouda//registration-config.json /home/amandapotts/git/arkouda//ServerModules.cfg /home/amandapotts/git/arkouda//src
chpl dependencies are missing - try make test-venv or make chapel-py-venv
registered  15  commands from  4  modules
chpl --print-passes    --ccflags="-DH5_USE_110_API" --using-attribute-toolname arkouda --no-checks --no-loop-invariant-code-motion --no-fast-followers --ccflags="-O0" -smemTrack=true -smemThreshold=1048576 --no-cache-remote --interleave-memory -I/home/amandapotts/anaconda3/envs/arkouda-dev/include -L/home/amandapotts/anaconda3/envs/arkouda-dev/lib --ldflags="-Wl,-rpath,/home/amandapotts/anaconda3/envs/arkouda-dev/lib" -I/home/amandapotts/git/arkouda//dep/libiconv-install/include -L/home/amandapotts/git/arkouda//dep/libiconv-install/lib --ldflags="-Wl,-rpath,/home/amandapotts/git/arkouda//dep/libiconv-install/lib" -lhdf5 -lhdf5_hl -lzmq -liconv -lidn2 -lparquet -larrow -sarkoudaVersion="\"v2024.10.02+46.gbe369ed9f\""  /home/amandapotts/git/arkouda//src/arkouda_server.chpl -M /home/amandapotts/git/arkouda//src/compat/eq-20  /home/amandapotts/git/arkouda//src/CheckpointMsg.chpl /home/amandapotts/git/arkouda//src/IndexingMsg.chpl /home/amandapotts/git/arkouda//src/nd_support/nd_array_stamps.chpl -sMaxArrayDims=1  -sOpenReaderLockingDefault=false -sOpenWriterLockingDefault=false  /home/amandapotts/git/arkouda//src/registry/Commands.chpl -I/home/amandapotts/git/arkouda//src/parquet -o arkouda_server
Timing for driver compilation phase
--------------
                            init :   0.001 seconds
             parseAndConvertUast :   1.065 seconds
                       checkUast :   0.085 seconds
                     readExternC :   0.004 seconds
          expandExternArrayCalls :   0.005 seconds
                         cleanup :   0.045 seconds
/home/amandapotts/git/arkouda//src/RadixSortLSD.chpl:25: error: Bad identifier in 'except' clause, no known 'isSorted' defined in 'ArkoudaSortCompat'
make: *** [Makefile:397: arkouda_server] Error 1
```

e-kayrakli replied:
If you can give this another try, that'd be great. I made a lot of changes in my current iteration and worked only with Chapel 2.3.

ajpotts commented on src/CheckpointMsg.chpl:
```diff
@@ -0,0 +1,188 @@
+    writeln("numTargetLocales ", numTargetLocales);
+
+    assert(numTargetLocales==1);
+
```
Why the `assert(numTargetLocales==1); `?

e-kayrakli replied:
I replaced asserts with custom errors thrown. They typically represent a "strange" issue with the logic, including things like trying to load a checkpoint saved with a different number of locales. We could/should relax those limitations in the near term.

ajpotts commented:
Thanks for putting this up!  It looks great.  I noted some comments as I was looking over it.  Really looking forward to this new feature!

e-kayrakli commented on src/CheckpointMsg.chpl:
```diff
@@ -0,0 +1,188 @@
+    for f in filenames {
+      mdWriter.writef("%i %s\n", f.numCodepoints, f);
+    }
+
```
done in the current revision

e-kayrakli commented:
@ajpotts -- I took another focused look at this and I think it is ready for another round of review. I believe I've addressed all your concerns and tested enough. Based on how you feel about it I am ready for it to be merged, but happy to address any other feedback. At a high-level, things happened since your first review:

- Added a `mode` argument to `save_checkpoint`: This controls whether an existing checkpoint should be overridden, or that should cause an error.
- Converted all metadata into JSON format
    - Added a `ArkoudaJSONCompat` module to be able to use relatively new features in the standard `JSON` module
- Added `LoadCheckpointError` and `SaveCheckpointError` types. They are thrown from several places where a checkpointing logic error occurred. Note that other IO etc errors can be propagated, too. IOW, checkpointing doesn't consume other errors to throw just these two errors.
    - There are no `assert` s or `try!` in the Chapel code as a result.
- Added some more `cpLogger` output for debugging purposes.
- `writeDistArrayToParquet` now also returns file sizes. This is necessary for us to be able to load distributed arrays that were checkpointed. I think that's something we can improve upon -- when reading Parquet, we need to pass sizes of each file to be read. In fact, these numbers can be calculated based on the destination arrays distribution. That could be done via an additional helper in the parquet interface, but not high priority.
- Beefed up the test to check some error conditions. These new tests generate "fake" checkpoints that have some sort of fault in them and asks the server to load to check the error message.
- I tested this on 8 locales with 8G of data.

e-kayrakli commented:
I fixed another issue, rebased on `master`. CI passed in an earlier run, I expect this one to pass.

1RyanK reviewed (approved):
Seems to work as expected (though I am also running 2.3)

ajpotts commented on arkouda/io.py:
```diff
@@ -2020,6 +2022,92 @@ def read(
+    # ret = _build_objects(rep)
+    return generic_msg(cmd="load_checkpoint", args={"name": name,
+                                                    "path": path})
+
```
There should probably be a function to clean up or delete a checkpoint, although that could be a separate ticket.

e-kayrakli replied:
https://github.com/Bears-R-Us/arkouda/pull/3915/commits/32332d28c742aace73ded30f6ac18e3658bdd2f8

ajpotts commented on tests/checkpoint_test.py:
```diff
@@ -0,0 +1,164 @@
+            loaded = ak.load_checkpoint(cp_name)
+
+            assert(arr[3] == 0)
+
```
Did you mean to also assert that `arr[2] == 2`, since you set it above?

ajpotts commented on arkouda/io.py:
```diff
@@ -2020,6 +2022,92 @@ def read(
 
 
+def save_checkpoint(name="", path=".akdata", mode: str = "override"):
+    """
```
I think it's OK to leave the default behavior as it is for now.  It would be easy to change later.

jaketrookman reviewed (approved):
Looks good and will be ready to be merged once the flake8 issues are fixed

e-kayrakli commented on arkouda/io.py:
```diff
@@ -2020,6 +2022,92 @@ def read(
+    See Also
+    ---------
+    save_checkpoint
+
```
https://github.com/Bears-R-Us/arkouda/pull/3915/commits/f5102b90f8194c5ee4b8991ebfdfe2a21a67c560 added one. But it is the same as `save_checkpoint`'s.

e-kayrakli commented on tests/checkpoint_test.py:
```diff
@@ -0,0 +1,164 @@
+        cp_name = ak.save_checkpoint()
+
+        expected_dir = '.akdata/{}'.format(cp_name)
+
```
Not in this particular test, but I will do that in the next one.

This particular one tests the default `save_checkpoint`, which will drop the checkpoint in `$(CWD)/.akdata`. The next test uses a custom path, which I will stick in a temp directory.
